python/ray/train/v2/_internal/execution/worker_group/state.py

```diff
@@ -0,0 +1,137 @@
+    """Ongoing state of an active worker group.
+
+    Attributes:
+        start_time: The time when the worker group was started.
+        workers: The workers in the worker group.
+            These should always be in sorted order by world rank.
+        placement_group: The placement group for the worker group.
+        sync_actor: The synchronization actor for the worker group.
+    """
+
+    start_time: float
+    placement_group: PlacementGroup
+    workers: List[Worker]
+    sync_actor: ActorHandle
+
+    def shutdown(self):
+        self._shutdown_workers(
+            self.workers,
+        )
+        self._shutdown_placement_group(self.placement_group)
+        self._shutdown_sync_actor(self.sync_actor)
+
+
+class WorkerGroupStateBuilder(_WorkerGroupShutdownMixin):
+    """Builder for WorkerGroupState.
+
+    Example usage:
+        ```python
+        builder = WorkerGroupStateBuilder()
+        builder.with_start_time(time_monotonic())
+        builder.with_placement_group(placement_group)
+        builder.with_workers(workers)
+        builder.with_sync_actor(sync_actor)
+        state = builder.build()
+
+        builder.shutdown(patience_s=10)
+        ```
+    """
+
+    def __init__(self):
+        self.start_time = None
+        self.placement_group = None
+        self.workers = None
+        self.sync_actor = None
+
+    def with_start_time(self, start_time: float) -> "WorkerGroupStateBuilder":
+        self.start_time = start_time
+        return self
+
+    def with_placement_group(
+        self, placement_group: PlacementGroup
+    ) -> "WorkerGroupStateBuilder":
+        self.placement_group = placement_group
+        return self
+
+    def with_workers(self, workers: List[Worker]) -> "WorkerGroupStateBuilder":
+        self.workers = workers
+        return self
+
+    def with_sync_actor(
+        self, sync_actor: SynchronizationActor
+    ) -> "WorkerGroupStateBuilder":
+        self.sync_actor = sync_actor
+        return self
+
+    def build(self) -> WorkerGroupState:
+        required_attrs = {
+            "start_time": self.start_time,
+            "placement_group": self.placement_group,
+            "workers": self.workers,
+            "sync_actor": self.sync_actor,
+        }
+        missing = [name for name, attr in required_attrs.items() if attr is None]
+        if missing:
+            raise ValueError(
+                f"Cannot build incomplete state. Missing: {', '.join(missing)}"
+            )
+        return WorkerGroupState(
+            start_time=self.start_time,
+            placement_group=self.placement_group,
+            workers=self.workers,
+            sync_actor=self.sync_actor,
+        )
+
+    def shutdown(self):
+        if self.workers:
+            self._shutdown_workers(self.workers)
+        if self.placement_group:
+            self._shutdown_placement_group(self.placement_group)
+        if self.sync_actor:
+            self._shutdown_sync_actor(self.sync_actor)
```

Comment:
should this also clear the references (set things to None)? otherwise we end up in a weird builder state.

Reply:
Ah I originally didn't do this since the Builder itself ends up getting de-referenced by the caller but that does make sense. WIll make this change.